File: lua/queen-attack/queen-attack.lua
return function(pos)
    assert(pos.row >= 0 and pos.row < 8 and pos.column >= 0 and pos.column < 8)
    function pos.can_attack(queen)
        if pos.row == queen.row or pos.column == queen.column then
            return true
        end
        return false
    end
    return pos
end

File: zig/queen-attack/queen_attack.zig
pub const QueenError = error{
    InitializationFailure,
};

pub const Queen = struct {
    row: i8,
    col: i8,

    pub fn init(row: i8, col: i8) QueenError!Queen {
        if (row < 0 or row >= 8 or col < 0 or col >= 8) {
            return error.InitializationFailure;
        }
        return .{ .row = row, .col = col };
    }

    pub fn canAttack(self: Queen, other: Queen) QueenError!bool {
        return self.row == other.row or
            self.col == other.col or
            self.row - other.row == self.col - other.col or
            self.row + self.col == other.row + other.col;
    }
};
